fix(oauth): Propagate error details to the client

sdass1918 opened the pull request:
Closes #13627

## Problem

Currently, when an OAuth identity provider returns an error to the client app (e.g., `error=access_denied`), these error details are lost during the redirect flow. This prevents the UI from displaying a meaningful warning to the user, leading to a confusing user experience.

## Solution

This PR implements the proposal in issue #13627 to propagate these error details to the client application. The changes include:

1.  **Passing Error Parameters:** The `error` and `error_description` query parameters from the OAuth callback are now passed to the `renderEndOfLoginResponse` function.
2.  **Storing in Config:** These error details are added to the client-side `config` object that is rendered in the HTML response.
3.  **Client-Side Handling:** In `end_of_redirect_response.js`, the code now checks for `config.error`. If an error is present, it is stored in `localStorage` using the keys `[storagePrefix]error` and `[storagePrefix]error_description`.

This allows developers to read these values from `localStorage` in their application and display an appropriate error message to the user.

## How to Test

1.  Configure an OAuth login service.
2.  Simulate a login failure that results in the provider redirecting with `error` and `error_description` query parameters (e.g., by denying access from the provider's consent screen).
3.  After the redirect, check the browser's `localStorage`.
4.  Confirm that `Meteor.oauth.error` and `Meteor.oauth.error_description` contain the values from the query parameters.

graemian committed:
Add instructions for enabling back gesture on iOS

ebroder committed:
Re-enable TypeScript self-test

The latest release of zodern:types is able to pull type declarations
from a Meteor checkout, if it's being used, instead of only from the
package repository (which may not contain the latest changes). This
means that we should be able to re-enable the TypeScript test, which
verifies that the TypeScript skeleton does, in fact, produce a project
which passes type checking.

(Re-enabling the test also picked up a few latent bugs in the type
declarations, so fix those as well)

zodern committed:
Add missing dependency on jquery

zodern committed:
Fix error generating xunit output

Grubba27 committed:
Merge branch 'devel' into fix/test-in-console

Grubba27 committed:
to version 1.2.7

StorytellerCZ committed:
Merge branch 'devel' into fix/test-in-console

nachocodoner committed:
Merge branch 'devel' into fix/test-in-console

mjcctech committed:
fix: displaying google maps in mobile application

dupontbertrand committed:
Update docs for upsert in ServiceConfiguration

Commit:
Update 3.forms-and-events.md

In the instructions to add onSubmit and onChange to TaskForm.jsx there was an unused import for TaskCollections. Removing this had no effect on the useability of the code in this step.

fredmaiaarantes committed:
Fix CLI and ecmascript links and add leonardoventurini:scss as the recommended scss package.

carlosarraes committed:
feat: add support for dots in minimongo key names

Minimongo now allows dots (.) in MongoDB key names, aligning with modern
MongoDB support.

Fixes #13552

sdass1918 committed:
fix(oauth): Propagate error details to the client

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/meteor/meteor?pullRequest=13985) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/meteor/meteor?pullRequest=13985) before we can accept your contribution.<br/>**12** out of **13** committers have signed the CLA.<br/><br/>:white_check_mark: nachocodoner<br/>:white_check_mark: graemian<br/>:white_check_mark: Grubba27<br/>:white_check_mark: zodern<br/>:white_check_mark: StorytellerCZ<br/>:white_check_mark: italojs<br/>:white_check_mark: fredmaiaarantes<br/>:white_check_mark: carlosarraes<br/>:white_check_mark: sdass1918<br/>:white_check_mark: mjcctech<br/>:white_check_mark: ebroder<br/>:white_check_mark: dupontbertrand<br/>:x: Jacob Gilbert<br/><hr/>**Jacob Gilbert** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/meteor/meteor?pullRequest=13985) it.</sub>

italojs commented on packages/oauth/end_of_redirect_response.js:
```diff
@@ -12,6 +12,15 @@
 
+  if (config.error) {
+    try {
+      localStorage[config.storagePrefix + "error"] = config.error;
```
Could we use session storage? I will avoid leaving a few pieces of garbage in the local storage.

italojs commented on packages/oauth/end_of_redirect_response.js:
```diff
@@ -12,6 +12,15 @@
+      localStorage[config.storagePrefix + "error"] = config.error;
+      localStorage[config.storagePrefix + "error_description"] = config.error_description;
+    } catch (err) {
+      console.log("Error storing OAuth error in localStorage", err);
```
Please do not use console.log since it's a thread blocker, use Log.warn/error instead

italojs reviewed (changes requested):
Missed a few test scenarios where:
- We can see the test not passing into the devel branch
- it's passing in your branch

sdass1918 committed:
fix(oauth): Address feedback and add regression test

sdass1918 commented:
Thanks for the great feedback @italojs! I've updated the PR to address all your points:

* Switched to `sessionStorage` to avoid leaving data behind.
* Replaced `console.log` with `Log.warn`.
* Added a new regression test to cover the error-handling scenario.

Let me know if there's anything else!

sdass1918 committed:
fix(oauth): Correct async handling in test

sdass1918 committed:
fix(oauth): Fixed the requested changes

italojs commented on packages/oauth/end_of_redirect_response.js:
```diff
@@ -12,6 +12,15 @@
+      sessionStorage[config.storagePrefix + "error"] = config.error;
+      sessionStorage[config.storagePrefix + "error_description"] = config.error_description;
+    } catch (err) {
+      Log.warn("Error storing OAuth error in sessionStorage", err);
```
well... I'm sorry that I didn't talk about it before, but I'm thinking here, why do we have a try-catch in a property assignment? I mean, it wont throw any error here

The file runs at the client, so we will always have the sessionStorage, the maximum we can do is address a fallback to config.storagePrefix
`sessionStorage[(config.storagePrefix || 'oauth_error')  + "error_description"] = ...`

I vote to remove the try catch from here and from line 6

sdass1918 committed:
refactor(oauth): Remove unnecessary try-catch block

sdass1918 commented:
Good point, thanks! I've removed the unnecessary try...catch block from both line 6 and line 17.

Grubba27 commented on packages/oauth/end_of_redirect_response.js:
```diff
@@ -4,12 +4,13 @@
-    } catch (err) {
-      // We can't do much else, but at least the redirects goes on.
-    }
+  }
```
could we keep this `try/catch` ?

sdass1918 replied:
Hi @Grubba27, thanks for the review!

I actually removed this `try/catch` based on earlier feedback from @italojs in this same PR, who suggested it was unnecessary.

I'm happy to add it back if that's the preferred approach! Could you and @italojs let me know what the consensus is?

Thanks!

sdass1918 replied:
I think the `try/catch` block is not really necessary here since we are only assigning a value to a variable @Grubba27 .

italojs replied:
We shouldn't use try/catch; it's just a variable assignment. If it throws an error, then node definitely has an issue.

italojs reviewed (changes requested):
The PR is good but I just missed a few tests

sdass1918 committed:
Merge branch 'devel' into fix-oauth-error-handling

sdass1918 commented:
Hi @italojs,

Thanks for the feedback & for catching that! I’ve just updated the branch to be in sync with the latest.

Please let me know if there’s anything else I should adjust (tests, formatting, docs…) and I’ll get on it.

Thanks again for your review!

nachocodoner committed:
update caching strategy in meteor-selftest-windows workflow

nachocodoner committed:
use Windows tar instead of Git tar in workflow

italojs committed:
Merge branch 'devel' into fix-oauth-error-handling

italojs commented:
dont forget the tests @sdass1918 ;)

italojs committed:
Merge branch 'devel' into fix-link-and-recommended-scss

italojs committed:
Merge branch 'devel' into patch-1

italojs committed:
Merge pull request #13428 from JacobKGilbert/patch-1

Update 3.forms-and-events.md

italojs committed:
Merge branch 'devel' into patch-2

italojs committed:
Merge pull request #13404 from dupontbertrand/patch-2

Update docs for upsert in ServiceConfiguration

italojs committed:
Merge branch 'devel' into fix-link-and-recommended-scss

italojs committed:
Merge pull request #13477 from meteor/fix-link-and-recommended-scss

docs: Fix links and add leonardoventurini:scss as recommended

nachocodoner committed:
Merge branch 'devel' into cache-windows-ci-check

nachocodoner committed:
re-run checks

nachocodoner committed:
re-run checks

nachocodoner committed:
clean

nachocodoner committed:
disable submodule update failure check in install.ps1

nachocodoner committed:
temp disable clean to verify cache speed

nachocodoner committed:
simplify cache paths in Windows self-test workflow

nachocodoner committed:
simplify cache paths in Windows self-test workflow

Grubba27 committed:
FEATURE: Remove `import React from 'react';` from skels

With our SWC changes, we can both remove this import line and add an example of swc configuration

nachocodoner committed:
skip temporary --get-ready

Grubba27 committed:
Merge pull request #14041 from meteor/feature/remove-react-import

FEATURE: Remove `import React from 'react';` from skels

nachocodoner committed:
refactor CI install script to adapt for CI environment

nachocodoner committed:
update workflow name to 'Windows Selftest'

nachocodoner committed:
refine Windows self-test workflow configuration

nachocodoner committed:
rename workflow file and update path references

nachocodoner committed:
fix path for windows-selftest.yml in workflow triggers

nachocodoner committed:
update cache key in Windows CI workflow

nachocodoner committed:
refactor dependency installation steps in Windows CI workflow

nachocodoner committed:
Merge pull request #14032 from meteor/cache-windows-ci-check

Windows CI check optimizations

nachocodoner committed:
Merge branch 'devel' into release-3.4

nachocodoner committed:
remove wildcard branch trigger from Windows CI workflow

nachocodoner committed:
Merge branch 'devel' into release-3.4

nachocodoner committed:
re-run checks

italojs committed:
Merge branch 'release-3.4.1' into fix/test-in-console

italojs committed:
Merge pull request #13000 from meteor/fix/test-in-console

Fix bugs with test-in-console

nachocodoner committed:
simplify cache key by removing package-lock.json hash

italojs committed:
Merge branch 'devel' into patch-1

italojs committed:
Merge pull request #12909 from graemian/patch-1

Add instructions for enabling back gesture on iOS

italojs committed:
merge branch

nachocodoner committed:
Merge branch 'devel' into release-3.4

italojs committed:
Merge branch 'release-3.4' into release-3.4.1

italojs committed:
test: make typescript template test async and await s.init()

sdass1918 committed:
Merge branch 'release-3.4.1' into fix-oauth-error-handling

nachocodoner commented:
This one is a duplicate of https://github.com/meteor/meteor/pull/13974, which came first. This PR is also in a weird state, showing not only the changes needed for the feature.

I will close it. Please let me know if this PR addresses something else, because I can't tell from the current diff with so many unrelated changes.
